lib/src/main/java/tech/relaycorp/awaladroid/messaging/IncomingMessage.kt

```diff
@@ -62,5 +76,46 @@ public class IncomingMessage internal constructor(
                 ack = ack
             )
         }
+
+        private suspend fun processPDAPath(
+            pdaPathSerialized: ByteArray,
+            senderEndpoint: ThirdPartyEndpoint,
+            recipientEndpoint: FirstPartyEndpoint,
+        ) {
+            if (senderEndpoint is PublicThirdPartyEndpoint) {
+                logger.info(
+                    "Ignoring PDA path from public endpoint ${senderEndpoint.privateAddress} " +
+                        "(${senderEndpoint.publicAddress})"
+                )
+                return
+            }
+            val pdaPath = try {
+                CertificationPath.deserialize(pdaPathSerialized)
+            } catch (exc: CertificationPathException) {
+                logger.log(
+                    Level.INFO,
+                    "Ignoring malformed PDA path for ${recipientEndpoint.privateAddress} " +
+                        "from ${senderEndpoint.privateAddress}",
+                    exc,
+                )
+                return
+            }
+
+            try {
+                (senderEndpoint as PrivateThirdPartyEndpoint).updatePDAPath(pdaPath)
+            } catch (exc: InvalidAuthorizationException) {
+                logger.log(
+                    Level.INFO,
+                    "Ignoring invalid PDA path for ${recipientEndpoint.privateAddress} " +
+                        "from ${senderEndpoint.privateAddress}",
+                    exc,
+                )
+                return
+            }
+            logger.info(
+                "Updated PDA path from ${senderEndpoint.privateAddress} for " +
+                    recipientEndpoint.privateAddress
+            )
+        }
```

Comment:
If this code got extracted to a separate class, I feel it would be easier to test.